boot/boot.go

```diff
@@ -455,31 +454,58 @@ func makeBootable20RunMode(model *asserts.Model, rootdir string, bootWith *Boota
 		return fmt.Errorf("cannot write modeenv: %v", err)
 	}
 
-	// get the ubuntu-boot bootloader
+	ubuntuBootPartition := filepath.Join(runMnt, "ubuntu-boot")
+
+	// get the ubuntu-boot grub and extract the kernel there
 	opts := &bootloader.Options{
 		// TODO:UC20: we use "recovery: true" here because on
 		// the partition the file layout of ubuntu-boot looks
 		// the same as ubuntu-seed.
 		// TODO:UC20: need a better name than recovery
 		Recovery: true,
+
+		// extract efi kernel assets to the ubuntu-boot partition
+		ExtractedRunKernelImage: true,
 	}
-	bl, err := bootloader.Find(filepath.Join(runMnt, "ubuntu-boot"), opts)
+	bl, err := bootloader.Find(ubuntuBootPartition, opts)
 	if err != nil {
 		return fmt.Errorf("internal error: cannot find run system bootloader: %v", err)
 	}
-	// TODO:UC20: using the UC16/18 grubenv style until we have
-	// a UC20 grub.cfg and corresponding snapd early boot
-	// code
+
+	// extract the kernel first, then mark kernel_status ready in case we get
+	// rebooted in between
+	kernelf, err := snap.Open(bootWith.KernelPath)
+	if err != nil {
+		return err
+	}
+
+	err = bl.ExtractKernelAssets(bootWith.Kernel, kernelf)
+	if err != nil {
+		return err
+	}
+
+	// TODO:UC20: support non uefi kernel assets
```

Comment:
to avoid the detail leakage here we probably want to do something like with `RecoveryAwareBootloader`, have an optional bootloader interface like:

```
type ExtractedRunKernelImageBootloader {
    EnableKernel(snap.PlaceInfo) error  // makes the symlink
    EnableTryKernel(snap.PlaceInfo) error  // makes the symlink
    DisableKernel()  error // removes the symlink
    DisableTryKernel() error // removes the symlink
}
```

this should probably be done in a follow up though

Reply:
Do you think this shold be done here or in a followup?

Reply:
@mvo5 @anonymouse64  a follow up is fine, but is probably needed to do UC20 variants of markSuccessful and setNext cleanly

Reply:
I agree with @pedronis, this is better served in a followup IMHO

Reply:
@anonymouse64 depending on details we might also need bootloader methods to read the snap name, revision out of those symlinks to be able to cross check things between them and the modeenv, to clean up things if we end up in situations were we partially wrote things given that we cannot write symlinks, grubenv and modeenv in one go